Fixes and updates

HasAndries opened the pull request:
- Added **strict mode** to allow rigid unit testing and spot unknown fields or empty selections (It will throw an error)
- Added field to ExecutionError to identify where an error occurred
- Make _lookup protected to allow extension of type loading
- Fixed enum Coerce not working in some cases
- Added subscriptionType placeholder to avoid distracting errors from GraphiQL

HasAndries commented:
We currently use the library with production code and would like to use your official nuget package again. Could you please include these fixes and additions with the next release?

joemcbride commented on src/GraphQL/Execution/ExecutionError.cs:
```diff
@@ -1,17 +1,21 @@
         }
 
-        public ExecutionError(string message, Exception innerException)
+        public ExecutionError(Field field, string message, Exception innerException)
```
I wouldn't want to include the whole field object here.  Also since it doesn't always apply, I wouldn't want to make it a required field in the constructor.  I would probably be fine with adding a `string` of `FieldName` as a optional field.

HasAndries replied:
I used the whole field because it can have an alias and didn't want to pollute the code with conditionals. I'll make it an optional parameter.

joemcbride commented on src/GraphQL/Types/Schema.cs:
```diff
@@ -25,7 +25,7 @@ public interface ISchema : IDisposable
     public class Schema : ISchema
     {
-        private readonly Lazy<GraphTypesLookup> _lookup;
+        protected readonly Lazy<GraphTypesLookup> _lookup;
```
I do want to have a way to provide additional types, though I'm not sure this is the way I want to support it.  I was originally thinking of having a constructor overload that took in an `IEnumerable<Type>` or some sort of loader.

HasAndries replied:
I will remove this change as I have removed the code my side that needed access to it.

joemcbride commented on src/GraphQL/Types/EnumerationGraphType.cs:
```diff
@@ -33,7 +33,8 @@ public void AddValue(EnumValue value)
             var found = Values.FirstOrDefault(v => v.Value.Equals(value));
-            return found != null ? found.Name : null;
+            if (found != null) return found.Name;
+            return Values.FirstOrDefault(v => v.Name.Equals(value))?.Value;
```
I don't want to accept this change without a use case and unit tests.  Is this how the reference javascript project works?  I'm not sure if it is or not.  I do know we potentially have some issues here, though I don't want to make changes without locking down the behavior.

HasAndries replied:
If memory serves me right, the original code works when mapping out to JSON, but it does not work for inbound conversion of enums. I'll add a test.

joemcbride commented:
Made some comments inline.  The other issue is the merge commit.  This branch needs to be rebased on the master branch vs. having a merge commit.

joemcbride commented on src/GraphQL/Execution/DocumentExecuter.cs:
```diff
@@ -633,7 +637,7 @@ public bool DoesFragmentConditionMatch(ExecutionContext context, IHaveFragmentTy
             }
 
-            if (conditionalType.Equals(type))
+            if (conditionalType.ToString() == type.ToString())
```
I'm fairly sure we want to use `Equals` vs. a `ToString` here.  Do you have a use case?  Unit tests to prove?

HasAndries replied:
I see this was fixed since I started using the library source, it used to be `if (conditionalType == type)`. I'll revert to the code in master.

joemcbride commented on src/GraphQL/Execution/DocumentExecuter.cs:
```diff
@@ -530,7 +534,7 @@ public object CoerceValue(ISchema schema, GraphType type, object input, Variable
                 fields = new Dictionary<string, Fields>();
             }
-
+            if (selections == null) throw new Exception("Selection cannot be null");
```
This should get handled by validation, which is in progress.

joemcbride commented on src/GraphQL/Execution/DocumentExecuter.cs:
```diff
@@ -29,10 +29,12 @@ public class DocumentExecuter : IDocumentExecuter
     {
         private readonly IDocumentBuilder _documentBuilder;
         private readonly IDocumentValidator _documentValidator;
+        public bool Strict { get; set; }
```
I can understand what you want here, though would Validation perhaps solve the problem you're looking to solve?  Is that just essentially validation?  I don't want to move too far away from the official spec on behavior.

HasAndries replied:
I didn't know about the DocumentValidator, it looks like it can serve the same pupose. I'll make the change.

joemcbride replied:
I have a number of updates to the GraphQL Language in https://github.com/graphql-dotnet/graphql-dotnet/pull/65 to help support validation.  There are a number of validation scenarios that will be super hard to do until that is merged in.  If you are able to add your specific validation scenarios to https://github.com/graphql-dotnet/graphql-dotnet/issues/10 then we can double check that they aren't already solved with the core validation rules provided by the spec.

HasAndries commented:
Thanks for all the comments, I'll work through them and provide unit tests where applicable. We don't use rebase in our normal workflow, so I'll need to read up on how to use it without messing up the repositpory.

HasAndries commented:
I've updated our code to support the standard nuget package without the changes from this fork. The only fix that I have on my side is to do with `EnumerationGraphType` and I will submit a pull request for that.
When an enum value is passed as part of an Input, the `Coerce` for `EnumerationGraphType` gets passed a string value, and the lookup on Values fails, because it does a lookup on the actual enum value. That is why I changed the `Coerce` to lookup on `Values` if not found:
```
public override object Coerce(object value)
{
    var found = Values.FirstOrDefault(v => v.Value.Equals(value));
    if (found != null) return found.Name;
    return Values.FirstOrDefault(v => v.Name.Equals(value))?.Value;
}
```
